File: rs-vm/src/value.rs
use std::rc::Rc;
use std::fmt;
use std::fmt::{Debug, Formatter};

use vm::CodeObject;
use builtins::NativeFnCode;

#[derive(Debug)]
pub enum Value {
    Int(isize),
    Bool(bool),
    Char(char),

    Symbol(Symbol),
    Keyword(Symbol),

    Tuple(Vec<ValueRef>),

    Record(Record),
    Singleton(Singleton),

    RecordType(FieldType),
    SingletonType(FieldlessType),
    AbstractType(FieldlessType),
    BuiltInType(FieldlessType),

    Fn(Closure),
    NativeFn(NativeFn),
    MultiFn(MultiFn)
}

pub type ValueRef = Rc<Value>;

#[derive(Debug)]
pub struct Symbol(Option<String>, String);

#[derive(Debug)]
pub struct Record {
    typ: ValueRef,
    vals: Vec<ValueRef>
}

#[derive(Debug)]
pub struct Singleton {
    typ: ValueRef
}

#[derive(Debug)]
pub struct FieldType {
    name: ValueRef,
    supertyp: ValueRef,
    field_names: Vec<String>
}

#[derive(Debug)]
pub struct FieldlessType {
    name: ValueRef,
    supertyp: ValueRef
}

#[derive(Debug)]
pub struct Closure {
    pub codeobj: Rc<CodeObject>,
    pub clovers: Rc<Vec<ValueRef>>
}

pub struct NativeFn {
    pub code: NativeFnCode
}

impl Debug for NativeFn {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        write!(f, "#<NativeFn @ {:p}>", self)
    }
}

#[derive(Debug)]
pub struct MultiFn {
    name: ValueRef,
    // methods: RefCell<HashMap<(Vec<TypeMatcher>, Option<TypeMatcher>), ValueRef>>
}

#[derive(Debug)]
enum TypeMatcher {
    Isa(ValueRef),
    Identical(ValueRef)
}

impl Value {
    pub fn unwrap_int(&self) -> isize {
        if let Value::Int(i) = *self {
            i
        } else {
            panic!()
        }
    }
    
    pub fn unwrap_bool(&self) -> bool {
        if let Value::Bool(b) = *self {
            b
        } else {
            panic!()
        }
    }

    pub fn unwrap_codeobj(&self) -> Rc<CodeObject> {
        if let Value::Fn(Closure { ref codeobj, .. }) = *self {
            codeobj.clone()
        } else {
            panic!()
        }
    }

    pub fn unwrap_clovers(&self) -> Rc<Vec<ValueRef>> {
        if let Value::Fn(Closure { ref clovers, .. }) = *self {
            clovers.clone()
        } else {
            panic!()
        }
    }
}

File: rpython/centring/value.py

class Value(object):
    pass

class Type(Value):
    def __init__(self, name):
        self._namesym = name

    def type(self):
        return Type._type

class Symbol(Value):
    def __init__(self, ns, name=None):
        if name is None:
            self._ns = None
            self._name = ns
        else:
            self._ns = ns
            self._name = name

Type._type = Type(Symbol(u"centring.lang", u"Type"))
Symbol._type = Type(Symbol(u"centring.lang", u"Symbol"))

class Int(Value):
    _type = Type(Symbol(u"centring.lang", u"Int"))
    
    def __init__(self, i):
        self._int = i

    def type(self):
        return Int._type

    def unwrap_int(self):
        return self._int

class Bool(Value):
    _type = Type(Symbol(u"centring.lang", u"Bool"))
    
    def __init__(self, b):
        self._bool = b

    def type(self):
        return Bool._type

    def unwrap_bool(self):
        return self._bool

class Record(Value):
    def __init__(self, type, field_vals):
        self._type = type
        self._fields = field_vals

    def type(self):
        return self._type

    def rec_ref(self, i):
        return self._fields[i]

